packages/artifact/src/internal/status-reporter.ts

```diff
@@ -0,0 +1,94 @@
+import {info} from '@actions/core'
+
+/**
+ * Status Reporter that displays information about the progress/status of an artifact that is being uploaded or downloaded
+ *
+ * Variable display time that can be adjusted using the displayFrequencyInMilliseconds variable
+ * The total status of the upload/download gets displayed according to this value
+ * If there is a large file that is being uploaded, extra information about the individual status can also be displayed using the updateLargeFileStatus function
+ */
+
+export class StatusReporter {
+  private totalNumberOfFilesToProcess = 0
+  private processedCount = 0
+  private displayFrequencyInMilliseconds: number
+  private largeFiles = new Map<string, string>()
+  private totalFileStatus: NodeJS.Timeout | undefined
+  private largeFileStatus: NodeJS.Timeout | undefined
+
+  constructor(displayFrequencyInMilliseconds: number) {
+    this.totalFileStatus = undefined
+    this.largeFileStatus = undefined
+    this.displayFrequencyInMilliseconds = displayFrequencyInMilliseconds
+  }
+
+  setTotalNumberOfFilesToProcess(fileTotal: number): void {
+    this.totalNumberOfFilesToProcess = fileTotal
+  }
+
+  start(): void {
+    const _this = this
```

Comment:
Is there a benefit to doing this? Why not use `this` everywhere?

Reply:
There is a interesting typescript rule that comes up which is why I had to set it up the way it is:

![image](https://user-images.githubusercontent.com/16109154/78556351-23e0d080-780f-11ea-91dd-7848429f31d1.png)

Reply:
Are you referring to the implicit any error or the outer value of this?

Either way, this seems way an odd way to work around it. Is this the recommended solution to that outer value this?

Reply:
It looks like [arrow functions](https://www.typescriptlang.org/docs/handbook/functions.html#this-and-arrow-functions) would solve this issue, it makes `this` behave like you'd expect (coming from a C# world).

Reply:
When I added this code in my previous PR, there was some stack overflow answer that recommended using `_this` which is why I went down that route (can't find the exact thread now).

Basically though, there are 3 ways that you can deal with `this` while inside another function. All of them work, each one is acceptable, just important to be consistent.

https://medium.com/@denny.headrick/this-bind-this-a-javascript-story-313ff84a12c8
https://github.com/airbnb/javascript/issues/81#issuecomment-23173941

I've changed everything to arrow functions, seems nicer.